apiserver/facades/agent/caasagent/caasagent_test.go

```diff
@@ -4,46 +4,223 @@
+	"github.com/juju/juju/environs/cloudspec"
+	"github.com/juju/juju/internal/errors"
 	coretesting "github.com/juju/juju/internal/testing"
+	"github.com/juju/juju/internal/uuid"
+	"github.com/juju/juju/rpc/params"
 )
 
 var _ = gc.Suite(&caasagentSuite{})
 
 type caasagentSuite struct {
 	coretesting.BaseSuite
 
-	resources  *common.Resources
-	authorizer *apiservertesting.FakeAuthorizer
+	registry facade.WatcherRegistry
+
+	modelUUID coremodel.UUID
+
+	modelService              *MockModelService
+	modelConfigService        *MockModelConfigService
+	controllerConfigService   *MockControllerConfigService
+	externalControllerService *MockExternalControllerService
+	controllerConfigState     *MockControllerConfigState
+	stubService               *MockStubService
+
+	facade *caasagent.FacadeV2
+	result cloudspec.CloudSpec
 }
 
 func (s *caasagentSuite) SetUpTest(c *gc.C) {
 	s.BaseSuite.SetUpTest(c)
 
-	s.resources = common.NewResources()
-	s.AddCleanup(func(_ *gc.C) { s.resources.StopAll() })
+	var err error
+	s.registry, err = registry.NewRegistry(clock.WallClock)
+	c.Assert(err, jc.ErrorIsNil)
 
-	s.authorizer = &apiservertesting.FakeAuthorizer{
-		Tag: names.NewMachineTag("0"),
+	s.modelUUID = modeltesting.GenModelUUID(c)
+
+	credential := cloud.NewCredential("auth-type", map[string]string{"k": "v"})
+	s.result = cloudspec.CloudSpec{
+		Type:             "type",
+		Name:             "name",
+		Region:           "region",
+		Endpoint:         "endpoint",
+		IdentityEndpoint: "identity-endpoint",
+		StorageEndpoint:  "storage-endpoint",
+		Credential:       &credential,
+		CACertificates:   []string{coretesting.CACert},
+		SkipTLSVerify:    true,
 	}
 }
 
+func (s *caasagentSuite) setupMocks(c *gc.C) *gomock.Controller {
+	ctrl := gomock.NewController(c)
+	s.stubService = NewMockStubService(ctrl)
+	s.modelService = NewMockModelService(ctrl)
+	s.controllerConfigService = NewMockControllerConfigService(ctrl)
+	s.modelConfigService = NewMockModelConfigService(ctrl)
+	s.externalControllerService = NewMockExternalControllerService(ctrl)
+	s.controllerConfigState = NewMockControllerConfigState(ctrl)
+
+	s.facade = caasagent.NewFacadeV2(
+		s.modelUUID, s.registry, s.controllerConfigService, s.modelConfigService,
+		s.externalControllerService, s.controllerConfigState, s.stubService, s.modelService)
+
+	return ctrl
+}
+
 func (s *caasagentSuite) TestPermission(c *gc.C) {
-	s.authorizer = &apiservertesting.FakeAuthorizer{
+	defer s.setupMocks(c).Finish()
+
+	authorizer := &apiservertesting.FakeAuthorizer{
 		Tag: names.NewApplicationTag("someapp"),
 	}
-	modelUUID := modeltesting.GenModelUUID(c)
 
-	_, err := caasagent.NewStateFacadeV2(facadetest.ModelContext{
-		Auth_:      s.authorizer,
-		ModelUUID_: modelUUID,
+	_, err := caasagent.NewFacadeV2AuthCheck(facadetest.ModelContext{
+		Auth_:      authorizer,
+		ModelUUID_: s.modelUUID,
 	})
 	c.Assert(err, gc.ErrorMatches, "permission denied")
 }
+
+func (s *caasagentSuite) TestCloudSpec(c *gc.C) {
```

Comment:
I would like to see a test here asking for the wrong model uuid just to confirm that everything blows up. Ideally this test should stop foot guns in the future.

Reply:
this is already tested
```
	result, err := s.facade.CloudSpec(
		context.Background(),
		params.Entities{Entities: []params.Entity{
			{names.NewModelTag(s.modelUUID.String()).String()},
			{otherModelTag.String()},
			{machineTag.String()},
		}},
	)	
```	
```
	}, {
		Error: &params.Error{
			Code:    params.CodeUnauthorized,
			Message: "permission denied",
		},
	}, {
```

Reply:
Ta